cocotb/decorators.py

```diff
@@ -93,6 +93,11 @@ def __init__(self, inst):
             self._natively_awaitable = True
         elif inspect.isgenerator(inst):
             self._natively_awaitable = False
+        elif inspect.iscoroutinefunction(inst):
+            raise TypeError(
+                "Coroutine function {} should be called prior to being "
+                "scheduled."
+                .format(inst))
```

Comment:
I'm thinking this isn't necessary either. This is never called directly and all the uses type check the arguments before calling the constructor. I second check that.

Reply:
Perhaps it's not, but it's consistent with how we already error on `isasyncgen` below, so I'm inclined to leave worrying about moving it to a later PR.

Reply:
Well ideally we would co-locate all the coroutine type checking stuff so we wouldn't have to keep repeating ourselves (inconsistently I'll add). I tried this in #1767, but I wasn't happy enough with the implementation to finish it out.